engine/runner/src/main/java/org/enso/runner/JavaFinder.java

```diff
@@ -0,0 +1,64 @@
+package org.enso.runner;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import org.enso.distribution.DistributionManager;
+import org.enso.distribution.Environment;
+
+/** Utility class that tries to find installed JDK on the system. */
+final class JavaFinder {
+  private JavaFinder() {}
+
+  /**
+   * Tries to find {@code java} executable on the system. If a system-wide JDK is not found, tries
+   * to find it in the {@link DistributionManager distribution} runtimes.
+   *
+   * @return null if cannot be found. Otherwise, returns the absolute path to the executable, or
+   *     simply {@code java} if it is on the {@code PATH}.
+   */
+  static String findJavaExecutable() {
+    var javaHome = System.getenv("JAVA_HOME");
+    if (javaHome != null) {
+      var java = new File(javaHome, "bin/java").getAbsoluteFile();
+      if (java.exists()) {
+        return java.getAbsolutePath();
+      }
+    }
+    if (isJavaOnPath()) {
+      return "java";
+    }
+    var javaInRuntime = findJavaExecutableInDistributionRuntimes();
+    if (javaInRuntime != null) {
+      return javaInRuntime.toAbsolutePath().toString();
+    }
+    return null;
+  }
```

Comment:
I think the precedence should actually look into distribution _first_ and only afterwards use system JVM.

The whole purpose of `ensoup` was to manage JVMs related to the engine versions, so that the engine is run with the right compatible JVM and not a random version that may be incompatible with it (I guess it used to be more important before Truffle 'got unchained').

I think ideally we should first to run the JVM version _required by the engine_ if it's found in the distribution. Then we can fallback to using the logic of searching `JAVA_HOME` and `Path`.

Ideally I'd also check the JVM version and log a warning if the version required by engine does not match the version that has been found.

Reply:
> JVMs related to the engine versions

This version is stored in the engine's manifest - see for example `built-distribution/enso-engine-0.0.0-dev-windows-amd64/enso-0.0.0-dev/manifest.yaml`.

It contains fields: `graal-vm-version` and `graal-java-version`.
You can see `lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/components/Manifest.scala` for the logic of loading the manifest and extracting the `GraalVMVersion` from it. 

We have a `GraalCEReleaseProvider` which translates this `GraalVMVersion` into a strategy for downloading it from GitHub - but you probably don't need this.

The function `graalDirectoryForVersion` in `RuntimeVersionManager` takes `GraalVMVersion` and gives you back the directory name under which this JVM version will be sitting in the `$ENSO_DATA_DIRECTORY/runtime` (installed by `ensoup`).

I guess you could factor out the `Manifest` and `graalDirectoryForVersion` utilities to a common package and you could use it to figure out the JVM version tied to the engine.

----

Btw. the manifest also contains a `jvm-options` field. I'd be very happy if we could keep the sometimes changing options in a single place.

Reply:
I guess if you want to check if the JVM version is correct you need to call `java --version` and somehow parse its output and compare with `graal-java-version`.

e.g. I get
```
openjdk 21.0.2 2024-01-16
OpenJDK Runtime Environment GraalVM CE 21.0.2+13.1 (build 21.0.2+13-jvmci-23.1-b30)
OpenJDK 64-Bit Server VM GraalVM CE 21.0.2+13.1 (build 21.0.2+13-jvmci-23.1-b30, mixed mode, sharing)
```

I guess I want to extract the second word from the first line. At least for now that seems to work. But we can update the logic whenever the version format of expected JVM changes - unexpected format just means it's not the JVM we expect :)

Reply:
Possibly better way than parsing output of `java -version` is to run a small program on the JVM:
```java
public static void main(String... args) {
  System.out.prinltn(System.getProperty("java.vm.version"));
  // etc.
}
```
and just read the output line by line without the need to parse words.

Reply:
Cool - that indeed sounds better, I didn't think of that